workflow/controller/operator.go

```diff
@@ -901,19 +902,31 @@ func (woc *wfOperationCtx) podReconciliation(ctx context.Context) error {
 				woc.updated = true
 			}
 			node := woc.wf.Status.Nodes[pod.ObjectMeta.Name]
+			match := true
+			if woc.execWf.Spec.PodGC != nil && woc.execWf.Spec.PodGC.LabelSelector != nil {
+				labelsMap, err := metav1.LabelSelectorAsMap(woc.execWf.Spec.PodGC.LabelSelector)
+				if err != nil {
+					woc.log.Warnf("Failed to parse label selector for pod GC: %s", woc.execWf.Spec.PodGC.LabelSelector)
+				}
+				labelSelector := labels.SelectorFromSet(labelsMap)
+				var podLabels labels.Set = pod.GetLabels()
+				match = labelSelector.Matches(podLabels)
+			}
 			if node.Fulfilled() && !node.IsDaemoned() {
 				if pod.GetLabels()[common.LabelKeyCompleted] == "true" {
 					return
 				}
-				woc.completedPods[pod.ObjectMeta.Name] = true
+				if match {
```

Comment:
I don't think we should changes the meaning of these arrays. Instead, do the selection at the point of deletion.

Reply:
Do you have other suggestions? Performing the selection here could avoid having to look up pod metadata since we have the metadata already during pod reconciliation. We only have pod names at the point of deletion.

Reply:
Good point.

What happens if we merge both into one `woc.completedPods` and instead of `map[string]bool` perhaps `map[string]podSummary`

```golang
type podSummary struct {
   succeeded bool
   match bool
}
```

???

Reply:
Great suggestion! I just merged these two arrays into a struct.